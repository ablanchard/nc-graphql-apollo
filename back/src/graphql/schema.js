import { makeExecutableSchema } from 'graphql-tools'
import types from './type'
import { resolvers as queryResolvers, queries } from './query'
import { resolvers as mutationResolvers, mutations } from './mutation'
import { resolvers as subscriptionResolvers, subscriptions} from './subscription'
import { merge } from 'lodash'

const typeDefs = [queries, mutations, subscriptions, types].join()

const resolvers = merge(queryResolvers, mutationResolvers, subscriptionResolvers)

export default makeExecutableSchema({
    typeDefs,
    resolvers
})