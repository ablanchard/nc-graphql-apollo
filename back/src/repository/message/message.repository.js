let id = 0
let messages = []

export const getMessages = function() {
    return messages
}

export const addMessage = function(message) {
    messages.push(message)
    return messages[messages.length - 1]
}

export const resetRepository = function() {
    messages = []
}

export const getNextId = function() {
    return id++
}