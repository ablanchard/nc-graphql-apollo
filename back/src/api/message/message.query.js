import { getMessages } from '../../service/message/message.service'

const MessageType = `
    type Message {
        id: ID!
        date: String!
        sender: Sender!
        content: String!
        localisation: String!
        status: String!
    }
`

const SenderType = `
    type Sender {
        pseudo: String!
        firstName: String!
        lastName: String!
    }
`

export const queries = `
    getMessages : [Message!]!
`

export const types = `
    ${MessageType}
    ${SenderType}
`

export const resolvers = {
    getMessages: () => {
        return getMessages()
    }
}