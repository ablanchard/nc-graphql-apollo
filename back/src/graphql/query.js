import { merge } from 'lodash'
import { queries as messageQueries, resolvers as messageResolvers } from '../api/message/message.query'

export const queries = `
    type Query {
        ${messageQueries}
    }
`
export const resolvers = {
    Query: merge(
        messageResolvers,
    )
}