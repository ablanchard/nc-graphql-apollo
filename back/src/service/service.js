import { init as initMessageSubscriber } from './message/message.subscriber'

export const initServices = async function() {
    await initMessageSubscriber()
}