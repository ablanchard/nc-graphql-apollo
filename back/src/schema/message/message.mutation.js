import { addMessage } from '../../service/message/message.service'

const MessageInput = `
    input MessageInput {
        content: String!
        localisation: String!
        status: String!
        sender: SenderInput!
    }
`

const SenderInput = `
    input SenderInput {
        pseudo: String!
        firstName: String!
        lastName: String!
    }
`

export const mutations = `
    saveMessage(message: MessageInput): Message
`

export const types = `
    ${MessageInput}
    ${SenderInput}
`

export const resolvers = {
    saveMessage: (_, {message}) => {
        return addMessage(message)
    }
}