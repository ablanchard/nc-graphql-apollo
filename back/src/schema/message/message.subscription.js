import { pubsub } from '../../service/websocket/websocket.service'

export const types = `
`

export const subscriptions = `
    subscribeMessages: Message
`

export const resolvers = {
    subscribeMessages: {
        resolve: (payload) => payload,
        subscribe: () => pubsub.asyncIterator(['subscribeMessages'])
    }
}