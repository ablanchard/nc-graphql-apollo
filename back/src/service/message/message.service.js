import { sleep } from '../utils.service'
import { eventBus } from './message.subscriber'
import * as messageRepository from '../../repository/message/message.repository'
import Config from 'config'


export const getMessages = function() {
    return messageRepository.getMessages()
}

export const addMessage = async function(message, latency = Config.latency) {
    await sleep(latency)
    const messageResult = messageRepository.addMessage({
        ...message,
        id: messageRepository.getNextId(),
        date: Date.now(),
        status: 'OK'
    })
    eventBus.next(messageResult)
    return messageResult
}

export const addCoinCoin = async function() {
    const coinMessage = {
        sender : {
            pseudo: 'Canard Man',
            firstName: 'Frédéric',
            lastName: 'Molas',
        },
        content: 'Coin Coin',
        localisation: 'Duckpound',
    }
    return await addMessage(coinMessage, 0)
}

export const isFromCanardMan = function(message) {
    return message.sender.pseudo === 'Canard Man'
}
