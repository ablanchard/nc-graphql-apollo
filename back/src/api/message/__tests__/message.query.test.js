jest.mock('../../../service/message/message.service')
import { getMessages } from '../../../service/message/message.service'
import { types, queries } from '../message.query'
import schema from '../../../graphql/schema'


describe('Schema of Message (Query)', () => {
    it('should be like snapshot (type)', () => {
        expect(types).toMatchSnapshot()
    })
    it('should be like snapshot (queries)', () => {
        expect(queries).toMatchSnapshot()
    })
})

describe('getMessages function', () => {
    const resolver = schema.getQueryType().getFields().getMessages.resolve
    let result
    beforeAll(async () => {
        getMessages.mockReturnValueOnce('RESULT_FAKE_MESSAGES')
        getMessages.mockClear()
        result = await resolver()
    })
    it('should call addMessage service with right parameter', () => {
        expect(getMessages).toHaveBeenCalled()
    })
    it('should return the result from getMessages service', () => {
        expect(result).toBe('RESULT_FAKE_MESSAGES')
    })
})