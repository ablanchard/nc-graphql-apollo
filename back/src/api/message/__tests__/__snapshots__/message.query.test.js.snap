// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema of Message (Query) should be like snapshot (queries) 1`] = `
"
    getMessages : [Message!]!
"
`;

exports[`Schema of Message (Query) should be like snapshot (type) 1`] = `
"
    
    type Message {
        id: ID!
        date: String!
        sender: Sender!
        content: String!
        localisation: String!
        status: String!
    }

    
    type Sender {
        pseudo: String!
        firstName: String!
        lastName: String!
    }

"
`;
